src/cargo/util/graph.rs

```diff
@@ -1,52 +1,219 @@
+        // the prefix we are resetting to had `age.len_nodes`, so remove all newer nodes
+        assert!(self.nodes.len() >= age.len_nodes);
+        while self.nodes.len() > age.len_nodes {
+            self.nodes.pop();
+        }
+
+        // the prefix we are resetting to had `age.len_edges`, so remove all links pointing to newer edges
+        for (_, lookup) in self.nodes.iter_mut() {
+            while lookup.len() >= 1
+                && lookup
+                    .get_index(lookup.len() - 1)
+                    .filter(|(_, idx)| idx >= &&age.len_edges)
+                    .is_some()
+            {
+                lookup.pop();
+            }
+        }
+
+        // the prefix we are resetting to had `age.len_edges`, so
+        assert!(self.edges.len() >= age.len_edges);
+        // remove all newer edges
+        let to_fix: Vec<EdgeIndex> = self
+            .edges
+            .drain(age.len_edges..)
+            .filter_map(|e| e.previous)
+            .filter(|idx| idx <= &age.len_edges)
+            .collect();
+
+        // fix references into the newer edges we are about to remove
+        for idx in to_fix {
+            self.edges[idx].next = None;
         }
+        assert_eq!(self.len(), age);
     }
 
     pub fn add(&mut self, node: N) {
-        self.nodes.entry(node).or_insert_with(im_rc::OrdMap::new);
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        self.nodes.entry(node).or_insert_with(IndexMap::new);
     }
 
-    pub fn link(&mut self, node: N, child: N) -> &mut E {
-        self.nodes
-            .entry(node)
-            .or_insert_with(im_rc::OrdMap::new)
-            .entry(child)
-            .or_insert_with(Default::default)
+    /// connect `node`to `child` with out associating any data.
+    /// Note that if this and `add_edge` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
+    pub fn link(&mut self, node: N, child: N) {
+        use indexmap::map::Entry;
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        match self
+            .nodes
+            .entry(node.clone())
+            .or_insert_with(IndexMap::new)
+            .entry(child.clone())
+        {
+            Entry::Vacant(entry) => {
+                // add the new edge, and link and fix the new link count
+                self.edges.push(EdgeLink {
+                    value: None,
+                    next: None,
+                    previous: None,
+                });
+                let edge_index: EdgeIndex = self.edges.len() - 1;
+                entry.insert(edge_index);
+            }
+            Entry::Occupied(_) => {
+                // this pare is already linked
+            }
+        };
+    }
+
+    /// connect `node`to `child` associating it with `edge`.
+    /// Note that if this and `link` are used on the same graph
+    ///      odd things may happen when `reset_to` is called.
```

Comment:
This seems like a bit of an odd restriction which may be a holdover from some old version of the resolver, would it be possible to only have either `link` or `add_edge`?

Reply:
`link` is only needed as a lockfile knows all the packages each package depends on but does not know the real `Dependency` objects ([hear](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/encode.rs#L265)). This could be a `Graph<_, ()>` but it needs to go into a `Resolve` so it needs to be  `Graph<_, Dependency>`. I am open to suggestions for how to work around this!

Having had a chance to sleep on it, I don't remember why I thought it would be a problem. I will just remove the "Note".